src/main/kotlin/com/team4099/robot2023/util/FieldFrameEstimator.kt

```diff
@@ -0,0 +1,224 @@
+        newVisionUpdates.add(visionUpdate)
+        newVisionUpdates.sortWith(VisionUpdate.compareDescStdDev)
+        updates[timestamp] =
+          PoseUpdate(prevUpdate.value.odomTRobot.exp(prevToVisionTwist), newVisionUpdates)
+      }
+    }
+
+    // Recalculate latest pose once
+    update()
+  }
+
+  /** Clears old data and calculates the latest pose. */
+  private fun update() {
+    // Clear old data and update base pose
+    // NOTE(parth): We need to maintain the history so that when vision updates come in, they have
+    // some buffer to
+    // interpolate within.
+    while (updates.size > 1 && updates.firstKey() < Clock.fpgaTime - HISTORY_LENGTH) {
+      val (_, value) = updates.pollFirstEntry()
+      baseOdometryTField = value.apply(baseOdometryTField, q)
+    }
+
+    // Update latest pose
+    odometryTField = baseOdometryTField
+    for (updateEntry in updates.entries) {
+      odometryTField = updateEntry.value.apply(odometryTField, q)
+    }
+
+    for (update in updates) {
+      if (update.value.visionUpdates.size > 0 && update.value.visionUpdates[0].fromVision) {
+        Logger.recordOutput("Vision/Buffer/Vision", update.key.inSeconds)
+
+        Logger.recordOutput(
+          "Vision/Buffer/VisionPose",
+          Pose2dWPILIB.struct,
+          update.value.visionUpdates[0].fieldTRobot.pose2d
+        )
+      } else {
+        Logger.recordOutput("Vision/Buffer/Drivetrain", update.key.inSeconds)
+      }
+    }
+  }
+
+  /**
+   * Represents a sequential update to a pose estimate, with a twist (drive movement) and list of
+   * vision updates.
+   */
+  private class PoseUpdate(val odomTRobot: Pose2d, val visionUpdates: ArrayList<VisionUpdate>) {
+    fun apply(previousOdomTField: Transform2d, q: Matrix<N3?, N1?>): Transform2d {
+      var currentOdomTField = previousOdomTField
+
+      // Apply vision updates
+      for (visionUpdate in visionUpdates) {
+        // Calculate Kalman gains based on std devs
+        // (https://github.com/wpilibsuite/allwpilib/blob/main/wpimath/src/main/java/edu/wpi/first/math/estimator/)
+        val visionK: Matrix<N3, N3> = Matrix(Nat.N3(), Nat.N3())
+        val r = DoubleArray(3)
+        for (i in 0..2) {
+          r[i] = visionUpdate.stdDevs.get(i, 0) * visionUpdate.stdDevs.get(i, 0)
+        }
+        for (row in 0..2) {
+          if (q.get(row, 0) === 0.0) {
+            visionK.set(row, row, 0.0)
+          } else {
+            visionK.set(
+              row, row, q.get(row, 0) / (q.get(row, 0) + Math.sqrt(q.get(row, 0) * r[row]))
+            )
+          }
+        }
+
+        // Calculate odom_T_field from this update's vision pose
+        val odomTVisionField =
+          odomTRobot.asTransform2d() + visionUpdate.fieldTRobot.asTransform2d().inverse()
+
+        // Calculate twist between current field frame transform and latest vision update
+        val odomTFieldDelta = odomTVisionField + currentOdomTField.inverse()
```

Comment:
visionField and field frame are not the same here, don't we want to do `odoTVisionField.inverse() + currentOdomTField` instead?

Reply:
also rename to visionTField

Reply:
yeah I think what we actually want here is `fieldTVisionField` because then that transform (in the exponential space) can be applied to `odomTField` to get the latest field frame transform